lib/spack/spack/container/writers/__init__.py

```diff
@@ -0,0 +1,153 @@
+        spack_version = self.container_config['base']['spack']
+        image, tag = build_info(run_image, spack_version)
+
+        Build = collections.namedtuple('Build', ['image', 'tag'])
+        return Build(image=image, tag=tag)
+
+    @tengine.context_property
+    def strip(self):
+        """Whether or not to strip binaries in the image"""
+        return self.container_config.get('strip', True)
+
+    @tengine.context_property
+    def paths(self):
+        """Important paths in the image"""
+        Paths = collections.namedtuple('Paths', [
+            'environment', 'store', 'view'
+        ])
+        return Paths(
+            environment='/opt/spack-environment',
+            store='/opt/software',
+            view='/opt/view'
+        )
+
+    @tengine.context_property
+    def manifest(self):
+        """The spack.yaml file that should be used in the image"""
+        import jsonschema
+        # Copy in the part of spack.yaml prescribed in the configuration file
+        manifest = copy.deepcopy(self.config['spack'])
+        manifest.pop('container')
+
+        # Ensure that a few paths are where they need to be
+        manifest.setdefault('config', syaml.syaml_dict())
+        manifest['config']['install_tree'] = self.paths.store
+        manifest['view'] = self.paths.view
+        manifest = {'spack': manifest}
+
+        # Validate the manifest file
+        jsonschema.validate(manifest, schema=spack.schema.env.schema)
+
+        return syaml.dump(manifest, default_flow_style=False).strip()
+
+    @tengine.context_property
+    def packages(self):
+        """Additional system packages that are needed at run-time."""
+        package_list = self.container_config.get('packages', None)
+        if not package_list:
+            return package_list
+
+        image = self.container_config['base']['image']
+        update, install, clean = package_info(image)
+        Packages = collections.namedtuple(
+            'Packages', ['update', 'install', 'list', 'clean']
+        )
+        return Packages(update=update, install=install,
+                        list=package_list, clean=clean)
+
+    @tengine.context_property
+    def extra_instructions(self):
+        Extras = collections.namedtuple('Extra', ['build', 'final'])
+        extras = self.container_config.get('extra_instructions', {})
+        build, final = extras.get('build', None), extras.get('final', None)
+        return Extras(build=build, final=final)
+
+    @tengine.context_property
+    def labels(self):
+        return self.container_config.get('labels', {})
+
+    def __call__(self):
+        """Returns the recipe as a string"""
+        env = tengine.make_environment()
+        t = env.get_template(self.template_name)
+        return t.render(**self.to_dict())
+
+
+# Import after function definition all the modules in this package,
+# so that registration of writers will happen automatically
+import spack.container.writers.singularity  # noqa
```

Comment:
In keeping with my earlier comment about the factory, I'd rather explicitly iterate over these in the code instead of having them magically register themselves unless there's a substantial performance or correctness reason not to.

Reply:
See my previous comment, it's just a personal preference to use decorators as they act as a marker on the code that will be pulled in here otherwise. If we're not using decorators we should do something like:
```python
import spack.container.writers.docker
import spack.container.writers.singularity

_writer_factory = {
  'docker': spack.container.writers.docker.DockerContext,
  'singularity': spack.container.writers.singularity.SingularityContext
}
```
or otherwise turn `writers` from a package to a single module with everything in it. I just prefer this layout for code organization since it seems more modular and more readable "locally" to me - you don't need to jump around files to understand that a`SingularityContext` is a writer for `singularity` since it has a decorator on top.

Another thing I'd like to note is that `spack.container` is designed around the idea that other parts of Spack will interact with it only through the names in its "root" namespace, while leaving everything deeper as an implementation detail. Right now those names are `validate` and `recipe`, so the complexity of the mechanism here should be reduced by knowing that the code we need to consider is just the one in the `spack/container` folder.

Let me know what are your thoughts after this explanation. If you still think this code is too complex or too hard to read we'll find a common ground to have a different layout.

Reply:
One option is doing something similar to what we do for `spack/modules`.